scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,221 @@
+use {
+    crate::handshake::{
+        shared::{LOGON_FAILURE, LOGON_SUCCESS, MAX_WORKERS},
+        ClientLogon, VERSION,
+    },
+    agave_scheduler_bindings::{
+        PackToWorkerMessage, ProgressMessage, TpuToPackMessage, WorkerToPackMessage,
+    },
+    nix::sys::socket::{self, ControlMessage, MsgFlags, UnixAddr},
+    std::{
+        fs::File,
+        io::{IoSlice, Read, Write},
+        os::{
+            fd::{AsRawFd, FromRawFd},
+            unix::net::{UnixListener, UnixStream},
+        },
+        path::Path,
+        time::{Duration, Instant},
+    },
+    thiserror::Error,
+};
+
+type ShaqError = shaq::error::Error;
+
+const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(1);
+
+pub struct Server {
+    listener: UnixListener,
+
+    buffer: [u8; 1024],
+    buffer_len: usize,
+}
+
+impl Server {
+    pub fn new(path: impl AsRef<Path>) -> Result<Self, std::io::Error> {
+        let listener = UnixListener::bind(path)?;
+
+        Ok(Self {
+            listener,
+            buffer: [0; 1024],
+            buffer_len: 0,
+        })
+    }
+
+    pub fn accept(&mut self) -> Result<Session, HandshakeError> {
+        // Clear the buffer.
+        self.buffer_len = 0;
+
+        // Wait for next stream.
+        let (mut stream, _) = self.listener.accept()?;
+        stream.set_read_timeout(Some(HANDSHAKE_TIMEOUT))?;
+
+        match self.handle_logon(&mut stream) {
+            Ok(session) => Ok(session),
+            Err(err) => {
+                let reason = err.to_string();
+                let reason_len = u8::try_from(reason.len()).unwrap_or(u8::MAX);
+
+                #[allow(
+                    clippy::arithmetic_side_effects,
+                    reason = "2 + u8::MAX fits in usize on all reasonable systems"
+                )]
```

Comment:
I'd just toss a `.wrapping_add()` instead of the longer explanation haha

Reply:
I realize, but it's shorter than `#![allow(clippy::arithmetic_side_effects]`, and in most cases is clear why it's okay.

Reply:
1bfbabd5945fcec5f5cd0662f3a04cdc64b6b52d